packages/playwright-core/src/server/frames.ts

```diff
@@ -603,48 +609,71 @@ export class Frame extends SdkObject {
 
   private async _goto(progress: Progress, url: string, options: types.GotoOptions): Promise<network.Response | null> {
     return this.raceNavigationAction(async () => {
-      const waitUntil = verifyLifecycle('waitUntil', options.waitUntil === undefined ? 'load' : options.waitUntil);
-      progress.log(`navigating to "${url}", waiting until "${waitUntil}"`);
-      const headers = this._page._state.extraHTTPHeaders || [];
-      const refererHeader = headers.find(h => h.name.toLowerCase() === 'referer');
-      let referer = refererHeader ? refererHeader.value : undefined;
-      if (options.referer !== undefined) {
-        if (referer !== undefined && referer !== options.referer)
-          throw new Error('"referer" is already specified as extra HTTP header');
-        referer = options.referer;
-      }
-      url = helper.completeUserURL(url);
+      while (true) {
```

Comment:
I don't feel good about while true in _goto (our core functionality) over a redirect in har (our p3 functionality). Can we do client-side redirect instead?